### PR TITLE
buildPythonPackage: add prefix to pname

### PR DESCRIPTION
**Pull request by FRidh**
The `pname` passed to `stdenv.mkDerivation` consists of the name prefix
as well as the `pname` passed to `stdenv.mkDerivation`.

This is needed in case we want to test for valid `pname` in
`stdenv.mkDerivation`.
https://github.com/NixOS/nixpkgs/pull/68620#issuecomment-605599680

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by FRidh** [comment]
This needs a testing.

**Comment by FRidh** [comment]
This change is backwards-incompatible and forces packages to pass in a `pname`.

File: pkgs/development/interpreters/python/mk-python-derivation.nix [hunk 1]
```diff
@@ -94,17 +93,17 @@
...
 let
   inherit (python) stdenv;
 
   self = toPythonModule (stdenv.mkDerivation ((builtins.removeAttrs attrs [
-    "disabled" "checkPhase" "checkInputs" "doCheck" "doInstallCheck" "dontWrapPythonPrograms" "catchConflicts" "format"
+    "pname" "disabled" "checkPhase" "checkInputs" "doCheck" "doInstallCheck" "dontWrapPythonPrograms" "catchConflicts" "format"
   ]) // {
 
-  name = namePrefix + name;
+  pname = namePrefix + attrs.pname;
```
Comment (FRidh) [inline pkgs/development/interpreters/python/mk-python-derivation.nix]: maybe better to check for `pname` and then prefix. Same with `name`.

**Review by jonringer — APPROVED** [review]

**Comment by jonringer** [comment]
still some kinks to work out though

```
error: attribute 'pname' missing, at /home/jon/.cache/nixpkgs-review/pr-83656/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:106:24
(use '--show-trace' to show detailed location information)
nix eval --json (import /nix/store/9ywlmbslln9va0i2kmbj9gz1bdc40a44-nixpkgs-review-2.2.0/lib/python3.7/site-packages/nixpkgs_review/nix/evalAttrs.nix /tmp/tmp1r2qufsq) failed to run, /tmp/tmp1r2qufsq was stored inspection
https://github.com/NixOS/nixpkgs/pull/83656 failed to build
```

**Pull request closed without merging**